### ISSUE TITLE
make sure access to cmsweb* services is on port 8443 in RestInteractions.py

### ISSUE THREAD

**Issue by belforte** [dmwm/CRABServer#6232]
see
https://github.com/dmwm/CRABClient/issues/4948

and authoritative reference in
https://hypernews.cern.ch/HyperNews/CMS/get/webInterfaces/1750.html

This needs to be in a new CRABServer tag for inclusion in CRABClient

**Comment by belforte** [comment]
Indeed this
https://github.com/dmwm/CRABServer/blob/f98b7b2e48f4d91ec39056f341713c58c71efb13/src/python/RESTInteractions.py#L34
is used all over the places also in CRABServer.
So it looks like all accesses to CRABServer REST are not using port 8443. The WMCore code
https://github.com/dmwm/WMCore/blob/b0e11b6987bd57ae156c09232fe4566456a0209c/src/python/WMCore/Services/Service.py#L122-L124
```python
        # setup port 8443 for cmsweb services
        if cfg_dict['endpoint'].startswith("https://cmsweb"):
            cfg_dict['endpoint'] = cfg_dict['endpoint'].replace('.cern.ch/', '.cern.ch:8443/', 1)
```
is only in effect for CRAB when accessing external CMSWEB services like DBS  (and PhEDEx while we had it).

We will need an extensive validation. Can't push this in as a critical fix. And when testing new cmsweb-auth cluster will stick with explicitely adding port 8443 in the restHostname in the config file.

**Comment by belforte** [comment]
this relates to https://gitlab.cern.ch/cms-http-group/doc/-/issues/237
see  https://hypernews.cern.ch/HyperNews/CMS/get/webInterfaces/1752.html
pasted below for convenience

```
*** Discussion title: Web Interface Discussions

Hi,
as a part of cmsweb migration campaign to switch from x509 to token based
authentication [1] we added a port number to our apache logs within recent
cmsweb upgrade. And now, we can perform an analysis of our tools which use
standard or machine based ports. Here is an example of entries corresponding to
different ports:

- standard port 443
[06/Nov/2020:01:00:00 +0100] cmsweb.cern.ch 2001:1458:d00:1a::7b "GET /couchdb/reqmgr_workload_cache/_design/ReqMgr/_view/byinputdataset?key=\"/WprimeToTB_plusSM_TToHad_M-3300_mixed_sqrt_13TeV-comphep/RunIIFall18pLHE-102X_upgrade2018_realistic_v11-v1/LHE\"&include_docs=true HTTP/1.1" 200 [data: 7840 in 33459 out 11859 body 6706 us ] [auth: TLSv1.2 DHE-RSA-AES128-GCM-SHA256 "/DC=ch/DC=cern/OU=computers/CN=unified/vocms0268.cern.ch/CN=1207901089" "-" ] [ref: "-" "-" ] 443

- machine port 8443
[06/Nov/2020:01:00:11 +0100] cmsweb.cern.ch 188.185.116.100 "GET /reqmgr2/data/request?name=pdmvserv_task_HIG-RunIIAutumn18NanoAODv7-03741__v1_T_201103_051130_1450 HTTP/1.1" 200 [data: 3292 in 17630 out 2155 body 54771 us ] [auth: TLSv1.2 ECDHE-RSA-AES128-GCM-SHA256 "/DC=ch/DC=cern/OU=computers/CN=wmagent/vocms0280.cern.ch" "-" ] [ref: "-" "WMCore.Services.Requests/v002" ] 8443

Please note that port number is shown at the end of the apache log record now.

I performed basic analysis of using last two days of apache logs to extract
relevant information about our usage of ports, see
/afs/cern.ch/user/v/valya/public/cmsweb/cmsweb-ports.sh

The results can be found in these two files
/afs/cern.ch/user/v/valya/public/cmsweb/ports-443.txt
/afs/cern.ch/user/v/valya/public/cmsweb/ports-8443.txt

Each file provides unique API end-point, DN, user agent and port. Therefore,
now we can easily identify clients which still rely on port 443 and start the
migration process of those clients to switch to port 8443 which will allow
to proceed with token authentication plan [1].

Here is an example how to look-up your favorite service and analyze its usage
of port 443 among different clients (user-agents):

echo "dbs phedex crab reqmgr t0" | tr ' ' '\n' | awk '{print "echo "$1"; cat /afs/cern.ch/user/v/valya/public/cmsweb/ports-443.txt | grep "$1" | wc -l"}' | /bin/sh
dbs 
632

phedex
2815

crab
2728

reqmgr
96433

t0
1397

Part of those are correct since we expect web clients to use standard port, but
others should be fixed, e.g.
/dbs/prod/global/DBSReader/blocks "/DC=ch/DC=cern/OU=computers/CN=unified/vocms0268.cern.ch/CN=1043260977" "DBSClient/3.5.1/" 443

I suggest that all service responsible will perform this basic analysis
and report on [1] which services/clients will require attention.

Best,
Valentin.

[1] https://gitlab.cern.ch/cms-http-group/doc/-/issues/237

-------------------------------------------------------------
Visit this CMS message (to reply or unsubscribe) at: 
https://hypernews.cern.ch/HyperNews/CMS/get/webInterfaces/1752.html

```

**Comment by belforte** [comment]
reference on CMSWEB side: https://its.cern.ch/jira/browse/CMSKUBERNETES-44

**Comment by vkuznet** [comment]
@belforte , any progress on this? We need to speed up this somehow.

**Comment by belforte** [comment]
Hi @vkuznet . No progress yet. But i've managed to close other high priority migrations.
So should be able to work on this next week. Hopefully will also be able to get help from @ddaina or @sharmaprajesh .

**Comment by belforte** [comment]
Let's put this at top of list then. And too bad for nanoAOD publication.

**Comment by belforte** [comment]
reference: https://github.com/dmwm/CRABServer/pull/6233

**Comment by vkuznet** [comment]
ok, thanks, will wait then till next week.

**Comment by belforte** [comment]
We need to test both CRABServer and CRABCache. The latter is still being accessed on port 443 by CRABClient.
So steps are:
- [ ] build those with the change in #6233 and deploy in cmsweb-testbed.
- [ ] configure client by hand to use port 8443 and test
- [ ] build a new client version where port 8443 is default and test (hopefully it will be enough to tag new CRABServer and build CRBAClient with it so that it pulls new RESTInteractions.py)

I am  for keeping the port # out of https://gitlab.cern.ch/crab3/CRAB3ServerConfig/blob/master/cmsweb-rest-config.json since client needs to access CRABServer in any case to get that information.

**Comment by belforte** [comment]
the commit done in here was insufficient. More work needed to make sure that port number is changed, see:
https://github.com/dmwm/CRABServer/issues/6354#issuecomment-763609273

**Comment by belforte** [comment]
Not as bad as I feared. Initial fix was too strict in looking for a resthost pointing to cmsweb.

**Comment by belforte** [comment]
still some problem with inserting port number:
```
				  File "/data/user/WMCore/src/python/WMCore/Services/pycurl_manager.py", line 252, in request
				    curl.perform()
				error: (3, "Port number ended with ':'
```

**Comment by amaltaro** [comment]
if problem is with pycurl_manager.py module, then I guess the WMCore services are also affected by this. Which means:
*) please do get back to us in case you managed to append this 8443 port to cmsweb like URLs
*) or ask for help ;-)

**Comment by belforte** [comment]
@amaltaro please be aware that CRAB has its own implementation of HTTPRequests... (not my idea of).
Anyhow, so far no worry, I think it is simply adding the port number twice because CRAB code has this original sin of  abusing of `self` and things that should be private class members specific to an instance are instead exported to the top layer and then down to everybody.  I was naive in first implementation. Looks to me that pycurl_manager does not change the port. Do you have plans to add 8443 in there by any chance ? It should be done in only one place, that's all.

**Comment by amaltaro** [comment]
The common place I've implemented it was under:
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/Services/Service.py#L123

but of course, it only works for CMSWEB services wrapper that are a sub-class of `Services`. In the end, having changes to the lower layer (http/curl) will be our best choice.

**Comment by belforte** [comment]
Yeah, I (CRAB) rely on that implementation of yours for access to DBS, the only remaining WMCore Service we depend on. But somehow I'd rather not try to make CRABServer another WMCore/Services instance, even if surely would have been a better choice if done right in the beginning. So if you keep it at that level, I have to do in CRABServer/RESTInterfaces.py and deal with the CRAB idiosyncrasies, if you move it to pycurl layer, I don't do anything. I am also fine with starting one way and changing later.
I should have this thing of mine running by tomorrow.

**Comment by belforte** [comment]
from my logbook:  here's a kibana view of FE logs showing port # once selecting crab-preprod-tw0* IP's https://monit-timber-cmsweb.cern.ch/kibana/goto/fe8e407537c6d0258fd26f530d41fad2

**Comment by belforte** [comment]
fixed via https://github.com/belforte/CRABServer/commit/0f1bbbcd00d965e39f384c5e57cf77c62972303c

**Comment by amaltaro** [comment]
If I read it right, you implemented it in a such a way that, if we remove that fix from the WMCore/Services module and move it elsewhere, it wouldn't break anything for CRAB. Is that correct?

**Comment by belforte** [comment]
yes and no. CRAB will pass to pycurl_manager urls like cmsweb-*.cern.ch:8443 , if you make changes in there, make sure not to add port number twice, or let's coordinate so I remove my hack.
I am starting to think that I should have simply added port # in the config. file and have the code always use what the config says !

**BUT** I think DBS Python Client adds internallu port 8443 since years. We need to be careful not to add it twice.

**Comment by belforte** [comment]
latest fix tested in preprod. Merged and made a new tag : v3.210120p1
Will test more using the docker image.

**Comment by belforte** [comment]
latest tag v3.210122 solves this in our testing in preprod. Will deploy to production and look with monitoring if there's some corner case that we missed.
